core/test/qa_thread_pool.cpp

```diff
@@ -82,39 +83,180 @@ const boost::ut::suite ThreadPoolTests = [] {
+        pool.waitUntilInitialised();
+
+        auto fut = pool.execute([]() -> int { throw std::runtime_error("expected failure"); });
+
+        expect(throws<std::runtime_error>([&] { (void)fut.get(); }));
+    };
+
+    "ThreadPool: recycled task count increases"_test = [] {
+        using namespace gr::thread_pool;
+
+        BasicThreadPool pool("RecycleTest", TaskType::IO_BOUND, 1U, 1U);
+        pool.waitUntilInitialised();
+
+        const auto before = pool.numTasksRecycled();
+
+        for (std::size_t i = 0; i < 5; ++i) {
+            std::atomic<bool> flag{false};
+            pool.execute([&] {
+                flag = true;
+                flag.notify_all();
+            });
+            flag.wait(false);
+        }
+
+        std::this_thread::sleep_for(std::chrono::milliseconds(10)); // let worker recycle
+
+        expect(gt(pool.numTasksRecycled(), before)); // conservatively ≥
+    };
+
+    "ThreadPool: setThreadBounds valid and invalid cases"_test = [] {
+        using namespace gr::thread_pool;
+
+        BasicThreadPool pool("BoundsTest", TaskType::CPU_BOUND, 2U, 4U);
+        pool.waitUntilInitialised();
+
+        expect(nothrow([&] { pool.setThreadBounds(1U, 8U); }));
+        expect(pool.minThreads() == 1U);
+        expect(pool.maxThreads() == 8U);
+
+        expect(throws<std::invalid_argument>([&] { pool.setThreadBounds(0U, 8U); }));
+        expect(throws<std::invalid_argument>([&] { pool.setThreadBounds(2U, 0U); }));
+        expect(throws<std::invalid_argument>([&] { pool.setThreadBounds(5U, 4U); }));
+
+        expect(nothrow([&] { pool.setThreadBounds(3U, 3U); }));
+        expect(pool.minThreads() == 3U);
+        expect(pool.maxThreads() == 3U);
+    };
+};
+
+const boost::ut::suite<"gr::thread_pool Manager"> ThreadPoolManager = [] {
+    using namespace boost::ut;
+    using namespace gr::thread_pool;
+
+    "Manager: default pools registration and retrieval"_test = [] {
+        auto cpu = Manager::defaultCpuPool();
+        auto io  = Manager::defaultIoPool();
+
+        expect(cpu->type() == TaskType::CPU_BOUND);
+        expect(io->type() == TaskType::IO_BOUND);
+        expect(cpu->device() == "CPU");
+        expect(io->device() == "CPU");
+        auto [min, max] = cpu->threadBounds();
+        expect(eq(min, cpu->minThreads()));
+        expect(eq(max, cpu->maxThreads()));
+
+        std::atomic<std::size_t> flag{0UZ};
+        cpu->execute([&] {
+            flag = 1UZ;
+            flag.notify_all();
+        });
+        flag.wait(0UZ);
+
+        expect(cpu->numThreads() >= 1U) << "needed to execute one task to spawn one thread";
+        expect(cpu->isShutdown() == false);
+    };
+
+    "Manager: custom pool registration and execution"_test = [] {
+        auto& manager = Manager::instance();
+
+        auto custom = std::make_shared<ThreadPoolWrapper>(std::make_unique<BasicThreadPool>("MyPool", TaskType::IO_BOUND, 1, 2), TaskType::IO_BOUND, "my_pool", "VirtualDevice");
```

Comment:
```suggestion
auto custom = std::make_shared<ThreadPoolWrapper>(std::make_unique<BasicThreadPool>("MyPool", TaskType::IO_BOUND, 1, 2), "VirtualDevice");
```
Could the Type (and probably also name) here be taken from the pool itself? Otherwise it's not that clear whether they have to be the same/different and which one is used where.